A/B test landing in the editor

mkevins opened the pull request:
#### Related PR:

https://github.com/wordpress-mobile/WordPress-iOS-Shared/pull/306

### Description

This PR adds a custom `GutenbergViewController` for editing the home page. This is used after the last step in site creation to bring the user directly into the editor after the new site is launched. It provides a modified UI (with a custom navigation bar), and a completion callback to launch quickstart when editing is finished.

This flow is conditioned on an ExPlat experiment named "wpios_land_in_the_editor_phase1_v3", and should leave the current behavior intact for the control variation.

#### Fixes tracks event regression

This PR also addresses a minor regression in which the tracks event `enhanced_site_creation_preview_ok_button_tapped` was not emitted at the end of the site creation flow for users who entered that path via the authentication screen.

#### Known Issue(s)

##### Animations

At the end of the existing user flow for the treatment variation, there is an issue where multiple dismiss animations occur before the quick start dialog is presented. It may be worth exploring ways to make this smoother (i.e. faster).

##### Potential for stale starter-site content

I did not encounter this, however, I foresee that this could cause potential confusion in the future: when Gutenberg is updated to somehow modify the format of a particular block that is used in the content of one of the starter-sites, the editor will parse this content when initializing, and automatically migrate the existing content to the new format. A minor issue is that this causes the content to have changed, prompting the user to save unsaved changes when they're continuing to complete the flow. For users who have not made any changes, such a prompt may be confusing, imo, especially if this is their first experience with the app / editor.

## To test

**Note**: To change the experiment variation between `Control` (default) and `Treatment` (experiment):

1. Browse to the Abacus experiment `wpios_land_in_the_editor_phase1_v3` (internal ref: `pbxNRc-1fJ-p2`)
2. Follow the instructions on the `Bookmarklet` tooltip to set the `Control` or `Treatment` variation for your user.
  * Note that this seems to only work with an A8C user account (separate test accounts may not work)

Alternatively you may use a proxy (e.g. Charlie) to intercept the call to the experiments endpoint (`/wpcom/v2/experiments/`) and return the desired variation ([detailed instructions in this previous experiment](https://github.com/wordpress-mobile/WordPress-iOS/pull/16129)). E.g.

```
{
	"variations": {
		"wpios_land_in_the_editor_phase1_v3": "treatment"
	},
	"ttl": 60
}
```

For each of the tests below, please also verify that the tracks event: `enhanced_site_creation_preview_ok_button_tapped` is emitted.

### Control: Login or New user

1. Login to the app with a user that has the `Control` variation assigned (check note above)
2. Press the **Create new site** option at the bottom
3. Select a design and a domain to create a site
4. Press **OK** on the next screen
5. **Verify** that you are presented with the `Want a little help to manage your site` dialog
6. **Verify** that you land on the **My Site** screen after your selection

<details>
<summary>Video</summary>


https://user-images.githubusercontent.com/8507675/151495092-fdc69c83-ebf7-4f61-9532-50fbf02c81b4.mp4
</details>

### Treatment: Login or New user

1. Login to the app with a user that has the `Treatment` variation assigned (check note above)
2. Press the **Create new site** option at the bottom
3. Select a design and a domain to create a site
4. Press **OK** on the next screen
5. **Verify** that you are presented with the home page editor
6. **Verify** that the `landing_editor_shown` analytics event is emitted
7. Press the **Continue** button and save any changes
8. **Verify** that you are presented with the `Want a little help to manage your site` dialog
9. **Verify** that you land on the **My Site** screen after your selection

<details>
<summary>Video</summary>


https://user-images.githubusercontent.com/8507675/151495124-1b25ec6b-1c1a-4cab-8e10-b1664079dd34.mp4
</details>

### Control: Existing or Logged in user

1. Login to the app with a user that has the Control variation assigned (check note above)
2. On the **My Site** screen press the site picker selector
3. Press the **Plus(+)** button at the top-right
4. Choose the **Create a WordPress.com site** option
5. Select a design and a domain to create a site
6. Press **OK** on the next screen
7. **Verify** that you are presented with the `Want a little help to manage your site` dialog
8. **Verify** that you land on the **My Site** screen after your selection

<details>
<summary>Video</summary>


https://user-images.githubusercontent.com/8507675/151495199-33b7f225-a9e8-47e6-bb98-595c1064384f.mp4
</details>

### Treatment: Existing or Logged in user

1. Login to the app with a user that has the `Treatment` variation assigned (check note above)
2. On the **My Site** screen press the site picker selector
3. Press the **Plus(+)** button at the top-right
4. Choose the **Create a WordPress.com site** option
5. Select a design and a domain to create a site
6. Press **OK** on the next screen
7. **Verify** that you are presented with the home page editor
8. **Verify** that the `landing_editor_shown` analytics event is emitted
9. Press the **Continue** button and save any changes
10. **Verify** that you are presented with the `Want a little help to manage your site` dialog
11. **Verify** that you land on the **My Site** screen after your selection

<details>
<summary>Video</summary>

https://user-images.githubusercontent.com/8507675/151495226-dd4abafb-aa84-449c-ac06-d61ed8be1e4d.mp4
</details>


## Regression Notes
1. Potential unintended areas of impact
Site creation flow

2. What I did to test those areas of impact (or what existing automated tests I relied on)

Manual testing of site creation flows
3. What automated tests I added (or what prevented me from doing so)

I began writing tests for the ExPlat A/B test, but was unable to cleanly stub the necessary dependency since the variation is provided via an enum getter (Swift enums cannot be inherited :disappointed: )

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [ ] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

mkevins committed:
Extract site assembly completion to helper class

This will be a good place to house logic for explat branching later on

mkevins committed:
Add homepage getter to `Blog`

mkevins committed:
Implement basic homepage editor

This uses the regular page editor for now, and will be customized in
later commits.

mkevins committed:
Implement EditHomepageViewController

This implements a view controller (and related collaborators) for
editing a homepage during the site creation flow.

mkevins committed:
Merge branch 'develop' into feature/land-in-the-editor

mkevins committed:
Implement homepage editing completion callback

mkevins committed:
Add explat condition for landing in the editor

mkevins committed:
Merge branch 'develop' into feature/land-in-the-editor

mkevins committed:
Fix whitespace issues

mkevins commented on WordPress/Classes/Utility/AB Testing/ABTest.swift:
```diff
@@ -2,6 +2,7 @@ import AutomatticTracks
 
 enum ABTest: String, CaseIterable {
     case unknown = "unknown"
+    case landInTheEditorPhase1 = "landInTheEditorPhase1"
```
This experiment name is just a suggestion for now, and open to change if desired.

antonis replied:
The naming looks good to me @mkevins!
@ob1-dev do you think this is a good name for an ExPlat experiment or there are conventions we should follow for this?

mkevins replied:
On further investigating, I think the string of the experiment name should be lowercase snake case. I'll push a change to correct this.

mkevins replied:
Updated with https://github.com/wordpress-mobile/WordPress-iOS/pull/17429/commits/0d03fa60f2030ba0789c1dfd583d99baa65d3504

mkevins commented on WordPress/Classes/ViewRelated/Gutenberg/GutenbergViewController.swift:
```diff
@@ -874,10 +884,19 @@ extension GutenbergViewController: GutenbergBridgeDelegate {
         }
     }
 
+    // Not ideal, but seems the least bad of the alternatives
+    @objc func continueFromHomepageEditing() {
+        fatalError("This method must be overriden by the extending class")
+    }
```
From the current patterns in place for code reuse, this was the best way I could manage reusing the abstraction here without resorting to much larger refactoring. Welcome alternatives if there is a better way to achieve a similar result though.

antonis replied:
The current approach looks reasonable to me @mkevins 👍 
It adds the desired functionality without making things too complex. Probably we can proceed with a larger refactoring after we collect enough data to consider the new experimental flow a success and we want to remove the old implementation.

mkevins committed:
Cleanup more whitespace and a TODO

mkevins committed:
Fix whitespace issue

antonis commented:
Great work @mkevins 👍 
I went through the draft PR and I think you implementation looks and behaves really good 🎉 
I've only added a few comments which are more of suggestions or request for feedback 🙇

antonis reviewed (commented):
Thank you for your work on this @mkevins 🙇 
The code LGTM and I think we can proceed with the final testing with an ExPlat experiment.
A potential enhancement would be to add some analytics to track the land on the editor flow. On [Android I added one new event](https://github.com/wordpress-mobile/WordPress-Android/pull/15627/commits/32246018aaba8f1d6a1e4dd0e64f790fc147e56c) for this.

mkevins committed:
Merge branch 'trunk' into feature/land-in-the-editor

This merges and resolves conflicts in `SiteAssemblyWizardContent` that
arose from other changes to the onboarding flow to improve quick start.
This should be refactored, if possible, to simplify the
implementation(s) of the onboarding code paths.

mkevins committed:
Merge branch 'trunk' into feature/land-in-the-editor

mkevins committed:
Fix trailing whitespace

mkevins committed:
Use snake-case for ExPlat experiment name

mkevins commented:
> Thank you for your work on this @mkevins bow The code LGTM and I think we can proceed with the final testing with an ExPlat experiment. A potential enhancement would be to add some analytics to track the land on the editor flow. On [Android I added one new event](https://github.com/wordpress-mobile/WordPress-Android/pull/15627/commits/32246018aaba8f1d6a1e4dd0e64f790fc147e56c) for this.

Good idea, I will add this as well! :+1:

antonis commented on WordPress/Classes/Utility/AB Testing/ABTest.swift:
```diff
@@ -2,6 +2,7 @@ import AutomatticTracks
 
 enum ABTest: String, CaseIterable {
     case unknown = "unknown"
+    case landInTheEditorPhase1 = "wpios_land_in_the_editor_phase1"
```
Hey @mkevins 👋 
I've created a new experiment (`wpios_land_in_the_editor_phase1_v3`) based on the feedback (ref: `pbxNRc-1fJ-p2`)

mkevins replied:
Ah, right! Good catch! I'll update this. :+1:

mkevins replied:
Updated with https://github.com/wordpress-mobile/WordPress-iOS/pull/17429/commits/d7cb0b0dd923ae4d0018c37e609479e705b5300e

mkevins committed:
Update ExPlat experiment name to latest version

mkevins committed:
Add land in the editor analytics and fix quick start issue

This adds an analytics event and also fixes an issue with the conflicts
with the quick start dialog. This can be refactored, and is only a
rudimentary step to get things working again.

mkevins committed:
Refactor to simplify experiment branching code

This refactors the implementation of the ExPlat experiment to land in
the editor so that the logic in the consumer can be close to the
original implementation.

mkevins committed:
Rename helper for landing in the editor

mkevins commented:
> > Thank you for your work on this @mkevins bow The code LGTM and I think we can proceed with the final testing with an ExPlat experiment. A potential enhancement would be to add some analytics to track the land on the editor flow. On [Android I added one new event](https://github.com/wordpress-mobile/WordPress-Android/pull/15627/commits/32246018aaba8f1d6a1e4dd0e64f790fc147e56c) for this.
> 
> Good idea, I will add this as well! +1

I've added an analytics event w/ https://github.com/wordpress-mobile/WordPress-iOS/pull/17429/commits/2d38fe3835132a44e630b85207b4fdaecb133850

mkevins commented on WordPress/Classes/ViewRelated/System/WPTabBarController+ShowTab.swift:
```diff
@@ -1,3 +1,4 @@
+import WordPressKit
```
Not sure why these changes appeared here. I will push another commit to fix this (this file should have no changes from `trunk` now).

mkevins replied:
Resolved w/ https://github.com/wordpress-mobile/WordPress-iOS/pull/17429/commits/ecff147783328f0cb1d2a448716ae003bff46b77

mkevins committed:
Fix erroneous changes from refactor

It seems a few lines were inadvertantly left while cleaning up a
refactor. This commit removes them.

mkevins committed:
Merge branch 'trunk' into feature/land-in-the-editor

antonis commented on WordPress/Classes/ViewRelated/Site Creation/FinalAssembly/LandInTheEditorHelper.swift:
```diff
@@ -0,0 +1,39 @@
+    // This seems to be necessary before casting `AbstractPost` to `Page`.
+    private static func fetchAllPages(for blog: Blog, success: @escaping PostServiceSyncSuccess, failure: @escaping PostServiceSyncFailure) {
+        let options = PostServiceSyncOptions()
+        options.number = 20
```
I wonder if the fetch limit is needed here (I see that it is [also used in pages list](https://github.com/wordpress-mobile/WordPress-iOS/blob/a08cc72e4bfb15d4fa88c06c549c16941fb2a64a/WordPress/Classes/ViewRelated/Pages/PageListViewController.swift#L265)) but I think the Android side does not use it.

mkevins replied:
If left out, I think the default for .com REST API is 20 anyway, and for WP API I believe it is 10, so this makes them the same. Otoh, it shouldn't make a difference until / unless one of the starter sites had more than 10 pages (and for the usage here, the additional condition that the home page was the 11th page or after).

**Edit:** It turns out, things are a bit more complex. For starters, when `options` is omitted, a _different_ default number gets specific (40) in `PostServiceRemoteREST` ([link](https://github.com/wordpress-mobile/WordPressKit-iOS/blob/0a573566b81c53007eedee5b53431b20bdc77883/WordPressKit/PostServiceRemoteREST.m#L77)). For some reason, though, when I tested this, I wound up in an infinite loop. Debugging further, it seems that omitting the options when `syncAll` is true leads to a loop because options is `nil` this line: https://github.com/wordpress-mobile/WordPress-iOS/blob/a678dadb73b13752e1306872cf8219be94e6b456/WordPress/Classes/Services/PostService.m#L210 
and more and more requests are made, since the condition is never "satisfied". Debugging also reveal that `loadedPosts` (at the time of running) was filled with 324 remote posts, each one with the same post id, so I think it actually made hundreds of requests (serially) each time adding the same single post to the collection.

I believe this indicates the presence of another bug when `options` is omitted, so for now, I think it's safest to leave this in.

antonis replied:
Thank you for the extra information @mkevins 🙇 
> I think it's safest to leave this in.
I agree with that 👍

antonis commented on WordPress/Classes/ViewRelated/Gutenberg/GutenbergViewController.swift:
```diff
@@ -7,6 +7,8 @@ import Kanvas
 class GutenbergViewController: UIViewController, PostEditor, FeaturedImageDelegate {
 
+
+
```
We can remove the extra lines here

mkevins replied:
Resolved w/ https://github.com/wordpress-mobile/WordPress-iOS/pull/17429/commits/1bd6ff44927f35a724ec2013374a851b3a4f7386

antonis reviewed (approved):
Great work @mkevins 👍 
I tested the app on an iPhone SE 2020 (iOS 14.7.1) and everything worked as expected. The code changes are consistent and overall LGTM 🎉

mkevins committed:
Remove unnecessary whitespace

mkevins committed:
Update WordPressShared reference to a tagged release

mkevins commented:
Thank you for reviewing and testing this Antonis!
